crates/wasmtime/src/compile.rs

```diff
@@ -643,6 +725,326 @@ the use case.
+                            Some(callee)
+                        } else {
+                            None
+                        }
+                    })
+                },
+            )?;
+
+            for (f, func) in layer.iter().zip(layer_outputs.drain(..)) {
+                debug_assert!(outputs[*f].is_none());
+                outputs[*f] = Some(func);
+            }
+        }
+
+        // Fan out in parallel again and finish compiling each function.
+        engine.run_maybe_parallel(outputs.into(), |output| {
+            let mut output = output.unwrap();
+            match &mut output.function {
+                CompiledFunction::Function(f) => inlining_compiler.finish_compiling(
+                    f,
+                    output.func_body.take(),
+                    &output.symbol,
+                )?,
+                #[cfg(feature = "component-model")]
+                CompiledFunction::AllCallFunc(f) => {
+                    debug_assert!(output.func_body.is_none());
+                    inlining_compiler.finish_compiling(&mut f.array_call, None, &output.symbol)?;
+                    inlining_compiler.finish_compiling(&mut f.wasm_call, None, &output.symbol)?;
+                }
+            };
+            Ok(output)
+        })
+    }
+
+    /// Implementation of our inlining heuristics.
+    ///
+    /// TODO: We should improve our heuristics:
+    ///
+    /// * One potentially promising hint that we don't currently make use of is
+    ///   how many times a function appears as the callee in call sites. For
+    ///   example, a function that appears in only a single call site, and does
+    ///   not otherwise escape, is often beneficial to inline regardless of its
+    ///   size (assuming we can then GC away the non-inlined version of the
+    ///   function, which we do not currently attempt to do).
+    ///
+    /// * Another potentially promising hint would be whether any of the call
+    ///   site's actual arguments are constants.
+    ///
+    /// * A general improvement would be removing the decision-tree style of
+    ///   control flow below and replacing it with (1) a pure estimated-benefit
+    ///   formula and (2) a benefit threshold. Whenever the estimated benefit
+    ///   reaches the threshold, we would inline the call. Both the formula and
+    ///   the threshold would be parameterized by tunables. This would
+    ///   effectively allow reprioritizing the relative importance of different
+    ///   hint sources, rather than being stuck with the sequence hard-coded in
+    ///   the decision tree below.
+    fn should_inline(
+        tunables: &Tunables,
+        caller_size: u32,
+        caller_module: StaticModuleIndex,
+        caller_def_func: DefinedFuncIndex,
+        caller_needs_gc_heap: bool,
+        callee_size: u32,
+        callee_module: StaticModuleIndex,
+        callee_def_func: DefinedFuncIndex,
+        callee_needs_gc_heap: bool,
+    ) -> bool {
+        log::trace!(
```

Comment:
Given the quantity of parameters here and how this is probably going to get tuned over time, how about packaging this all up in `struct InlineDecisionParams { .. }` or something like that? That way there's named keys and no need to worry about getting things in the wrong order by accident.